File: programmers/02.stack,queue/02.올바른 괄호/python/01.py
def solution(s):
    cnt = 0
    for item in s:
        if item == "(":
            cnt+=1
        else:
            cnt-=1
            if cnt < 0: 
                return False

    return True if cnt == 0 else False

# test
print(solution("()()"))
print(solution("(())()"))
print(solution(")()("))
print(solution("(()("))

File: programmers/02.stack,queue/02.올바른 괄호/JS/01.js
function solution(s) {
    let ans = true;
    const arr = s.split("");
    let openNum = 0;
    let closeNum = 0;
    arr.forEach((e) => {
        if (e === "(") {
            openNum++;
        } else {
            closeNum++;
            if (closeNum > openNum) ans = false;
        }
    });

    return ans ? (openNum === closeNum ? true : false) : false;
}

// test
console.log(solution("()()"));
console.log(solution("(())()"));
console.log(solution(")()("));
console.log(solution("(()("));
